### ISSUE TITLE
Rewrite 3rd party JS code to enforce that parent + top references are always correct

### ISSUE THREAD

**Issue by whites11** [cypress-io/cypress#1467]
### Current behavior:

I setup a very simple test that searches for a link in a web page (given an id) and clicks it. Than it checks the content of another element to check it's correct.
What happens is that selectors are found but the test fails. In fact the web page is behaving correctly.
I can see an error in the inspector:

`Uncaught SyntaxError: Unexpected token <`

And another one:

`Unhandled rejection TypeError: Cannot read property 'currentRetry' of undefined`

### Desired behavior:

The test to pass.

### How to reproduce:
Just run the test below.

#### Test code:

```js
describe('Check properties worldwide page', function() {
  beforeEach(function () {
    cy.visit('https://beta.hostelsclub.com/en/hostels');
  })

  it('Opens the correct country page', function () {
    cy.get('a#usa').click();
    cy.get('header.header-country .intro h1').contains('Best hostels in USA');
  })
});
```

### Additional Info (images, stack traces, etc)

![err2](https://user-images.githubusercontent.com/868430/37479569-40e69cac-287d-11e8-998d-3126b9bdeb94.png)

- Operating System: Arch Linux 
- Cypress Version: 2.1.0
- Browser Version: Electron 59

**Comment by jennifer-shehane** [comment]
Usually when there is an error like this, `Uncaught SyntaxError: Unexpected token <`, my first suspect is that a Network request is responding with HTML when it is expected that it should respond with non-HTML data. 

And indeed, there is a Network Request on your end for a `default-hostels.jpg` returning html content. I would fix this response on your end to respond with an image and maybe [write some logic to handle `onerror` for images](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/image#a-onerror).

![screen shot 2018-03-15 at 3 09 09 pm](https://user-images.githubusercontent.com/1271364/37485822-e916d1c8-2862-11e8-98b5-fdf26faeb415.png)

I guess there is the unanswered question of why this error appears this way instead of indicating that there was an error in your application code. 🤔

**Comment by whites11** [comment]
Hi Jennifer, thanks for your answer.
Your was a good guess, but apparently that is not the problem.
I removed the reference to the broken image from the web page but the problem still occours.

Is it possible to know which request is actually causing the error? I had a look at all of them and they look all fine to me now.

**Comment by whites11** [comment]
Ok thanks for the update. Any idea on how I can prevent the error (workaround) or why the problem suddently appeared while my tests were running fine since yesterday morning?

**Comment by brian-mann** [comment]
We investigated the problem - and its being caused exclusively by `sidecar-loader.js` which is 3rd party code.

The reason is partly Cypress' fault - when the `sidecar-loader.js` evaluates within Cypress, it will receive a different `window.document` reference - which then causes failures in its logic.

When it's evaluated outside of Cypress the `window.document` references are correct. This is in the same similar vein as what we do to prevent obstructive code from preventing Cypress from working - and we think the solution is that we can add additional checks to force javascript code to evaluate correctly (as if it's not running in Cypress).

### Workaround

You can simply workaround this problem by blacklisting the 3rd party sidecar code from loading, which will instantly make everything work. To do that, [read about blacklisting](https://docs.cypress.io/guides/references/configuration.html#blacklistHosts). You'll need to blacklist `sidecar.hoperator.com`

Here are my notes / thoughts just for internal reference on rewriting 3rd party JS code to enforce that `parent` + `top` references are always correct.

```
["parent"] ['parent] .parent getters => (window.parent && window.parent.__Cypress__ ? window : window.parent)
```

**Comment by whites11** [comment]
Hi Brian.
I confirm the workaround works.
Thanks a lot for your help guys, I was running out of options!

**Comment by matthewshirley** [comment]
I believe I have reproduced similar behavior with an internal application. I cannot block the 3rd party JS code that is being evaluated. Is there another workaround I can use?

**Comment by sensuraj92** [comment]
Any update on this issue please, i am also facing same issue, please suggest if we have any workaround

**Comment by flotwig** [comment]
This fix is available starting in [`4.6.0`](https://docs-staging.cypress.io/guides/references/changelog.html#4-6-0) as an [experiment](https://on.cypress.io/experiments) which you can access by setting this config option in your `cypress.json` or elsewhere:

```json
{
	"experimentalSourceRewriting": true
}
```

Please use a new issue if you are still running into issues that are not solved by this change.
